blockchain/wallet.go

```diff
@@ -147,9 +156,82 @@ func (s *Signature) Marshal() []byte {
 	return buf
 }
 
-// NewWallet produces a new Wallet that can sign transactionsand has a
+// NewWallet produces a new Wallet that can sign transactions and has a
 // public Address.
-func NewWallet() Wallet {
+func NewWallet() *Wallet {
 	priv, _ := ecdsa.GenerateKey(curve, crand.Reader)
-	return (*wallet)(priv)
+	return &Wallet{
+		PrivateKey: priv,
+		Balance:    0,
+	}
+}
+
+// SetAllPending appends transactions to the pending set of transactions.
+func (w *Wallet) SetAllPending(txns []*Transaction) {
+	for _, t := range txns {
+		w.SetPending(t)
+	}
+}
+
+// SetPending appends one transaction to the pending set of transaction
+// if the wallet effective balance is high enough to accomodate.
+func (w *Wallet) SetPending(txn *Transaction) error {
+	bal := w.GetEffectiveBalance()
+	spend := txn.GetTotalOutput()
+	if bal >= spend {
+		w.PendingTxns = append(w.PendingTxns, txn)
+	} else {
+		msg := fmt.Sprintf("wallet balance is too low %v < %v", bal, spend)
+		return errors.New(msg)
+	}
+	return nil
+}
+
+// DropAllPending drops pending transactions if they apper in txns.
+func (w *Wallet) DropAllPending(txns []*Transaction) {
+	for _, t := range txns {
+		if p, i := w.IsPending(t); p {
+			w.DropPending(i)
+		}
+	}
+}
+
+// DropPending a single pending transaction by index in the pending list.
+func (w *Wallet) DropPending(i int) {
+	if i < len(w.PendingTxns) && i >= 0 {
+		log.Info("dropping transaction with hash %s", w.PendingTxns[i].Input.Hash)
+		w.PendingTxns = append(w.PendingTxns[:i], w.PendingTxns[i+1:]...)
+	}
+}
+
+// IsPending returns true if the transaction exists in the pending list.
+// If true, it also returns the integer index of the transaction.
+func (w *Wallet) IsPending(txn *Transaction) (bool, int) {
+	for i, t := range w.PendingTxns {
+		if t.Input.Hash == txn.Input.Hash {
+			return true, i
+		}
+	}
+	return false, -1
+}
+
+// GetEffectiveBalance returns the wallet balance less the sum of the pending
+// transactions in the wallet.
+func (w *Wallet) GetEffectiveBalance() uint64 {
+	r := w.Balance
+	for _, t := range w.PendingTxns {
+		r -= t.GetTotalOutput()
```

Comment:
Is this mathematically correct? Won't part of the output to each transaction go back to the sender, meaning the total output is not actually how much you're spending?